### PR TITLE
fix: fix issue with kwargs while calling subroutine

### PR DESCRIPTION
**Pull request by atharva-satpute**
*Issue #11 :*

*Description of changes:*

- Convert kwargs to args to support keyword args while calling the subroutine function

*Testing done:* Yes

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/autoqasm/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/autoqasm/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/autoqasm/blob/main/README.md) and [API docs](https://github.com/amazon-braket/autoqasm/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/unit_tests/autoqasm/test_api.py [hunk 1]
```diff
@@ -1240,3 +1240,51 @@ def main():
...
+    def main():
+        test(a=0, b=1)
+
+    expected = """OPENQASM 3.0;
+def test(int[32] a, int[32] b) {
+    h __qubits__[a];
+    h __qubits__[b];
+}
+qubit[2] __qubits__;
+test();"""
```
Comment (laurencap) [inline test/unit_tests/autoqasm/test_api.py:1263]: I missed this on my first review. The correct output here would be `test(0, 1)`

**Review by laurencap — CHANGES_REQUESTED** [review]

File: test/unit_tests/autoqasm/test_api.py (hunk 1 above)
Comment (atharva-satpute) [reply test/unit_tests/autoqasm/test_api.py:1263]: I don't see this when I run the code locally

Comment (atharva-satpute) [reply test/unit_tests/autoqasm/test_api.py]: Reason for the error in the output: `args` were not updated after binding the `args` and `kwargs` to the function signature.
I have updated the code.

File: test/unit_tests/autoqasm/test_api.py [hunk 2]
```diff
@@ -1240,3 +1240,51 @@ def main():
...
+    """Test that subroutine calls works with keyword arguments"""
+
+    @aq.subroutine
+    def test(a: int, b: int) -> None:
+        aq.instructions.h(a)
+        aq.instructions.h(b)
+
+    @aq.main(num_qubits=2)
+    def main():
+        test(a=0, b=1)
```
[the remark this answers is not in the document]
Comment (rmshaffer) [reply test/unit_tests/autoqasm/test_api.py:1255]: I agree, this would be a good test case to add.

[the remark this answers is not in the document]
Comment (atharva-satpute) [reply test/unit_tests/autoqasm/test_api.py:1255]: Sure, I'll add the test case

[the remark this answers is not in the document]
Comment (atharva-satpute) [reply test/unit_tests/autoqasm/test_api.py:1255]: I have added the test case

**Review by laurencap — APPROVED** [review]

File: test/unit_tests/autoqasm/test_api.py [hunk 3]
```diff
@@ -1240,3 +1240,73 @@ def main():
...
+    """
+
+    @aq.subroutine
+    def test(a: int, b: int) -> None:
+        aq.instructions.h(a)
+        aq.instructions.h(b)
+
+    @aq.main(num_qubits=2)
+    def main():
+        test(0, b=1)
```
[the remark this answers is not in the document]
Comment (atharva-satpute) [reply test/unit_tests/autoqasm/test_api.py:1280]: Sure

[the remark this answers is not in the document]
Comment (atharva-satpute) [reply test/unit_tests/autoqasm/test_api.py]: Combined the test cases

**Review by rmshaffer — APPROVED** [review]

**Pull request merged**